### PR TITLE
Lesson 05 Assignment Files

### PR DESCRIPTION
**Pull request by rfelts**
Here are the lesson 05 assignment files. 

I did make a comment on line 140 of the mailroom file regarding trying the convert the for loop that begins on line 142 into a Comprehension. But the Comprehension kept overwriting my current value instead of appending.

### CODE REVIEW

File: students/rfelts/lesson05/mailroom.py [hunk 1]
```diff
@@ -0,0 +1,200 @@
...
+    for row in temp_data:
+        print(f"{row[0]:<{name_padding}}  ${row[1]:>{donation_padding},.2f} {row[2]:>11d} {row[3]:>{donation_padding},.2f}")
+
+
+def send_to_everyone():
+    """
+    Write a file for each donor containing the thank you message.
+    """
+
+    import datetime
```
[the remark this answers is not in the document]
Comment (rfelts) [reply students/rfelts/lesson05/mailroom.py]: Moved to the top of the file in the new PR

File: students/rfelts/lesson05/mailroom.py [hunk 2]
```diff
@@ -0,0 +1,200 @@
...
+
+def add_donation():
+    """
+    Prompt for a donor name. If the user enters a name not in the current donor list, add that name to the data
+    structure along with the donation amount. If the user exits just update the donation history. Then compose a thank
+    you email.
+    """
+
+    prompt_string = "\nPlease enter the donor's name: "
+    donor_name = input(prompt_string)
```
[the remark this answers is not in the document]
Comment (rfelts) [reply students/rfelts/lesson05/mailroom.py]: Fixed in latest PR

File: students/rfelts/lesson05/mailroom.py [hunk 3]
```diff
@@ -0,0 +1,200 @@
...
+    response = input(prompt_string)
+
+    # Verify the value entered is numeric if not prompt the user for a new value
+    while True:
+        try:
+            response = float(response)
+            return response
+        except ValueError:
+            print("The value enter was invalid.")
+            response = input(prompt_string)
```
[the remark this answers is not in the document]
Comment (rfelts) [reply students/rfelts/lesson05/mailroom.py]: Updated the code in latest PR

File: students/rfelts/lesson05/mailroom.py [hunk 4]
```diff
@@ -0,0 +1,200 @@
...
+           "{last_donation:<,.2f}.\nYou're total donations to date are now: {total_donations:<,.2f}.".format(**donor_info))
+
+
+def exit_menu():
+    """
+    Exits the current menu or program
+    :return - a String to exit
+    """
+
+    return "exit"
```
[the remark this answers is not in the document]
Comment (rfelts) [reply students/rfelts/lesson05/mailroom.py]: I use this method to exit the sub menu as well as the main menu. So mapping `sys.exit()` to this call will close the program when trying to exit the sub menu. If using `sys.exit()` is the preferred way to exit the main menu then I can create an exit program method and map it to that method.

File: students/rfelts/lesson05/mailroom.py [hunk 5]
```diff
@@ -0,0 +1,200 @@
...
+
+    import datetime
+    now = datetime.datetime.now()
+
+    donor_list = create_donor_list()
+
+    for donor in donor_list:
+        f = open(donor["name"] + '_' + str(now.month) + '_' + str(now.year) + '.txt', 'w')
+        f.write(compose_message(donor))
+        f.close()
```
[the remark this answers is not in the document]
Comment (rfelts) [reply students/rfelts/lesson05/mailroom.py]: Fixed in PR 113

File: students/rfelts/lesson05/mailroom.py [hunk 6]
```diff
@@ -0,0 +1,200 @@
...
+    if not any(s.lower() == donor_name.lower() for s in donor_list):
+        # Add the new donor and their donation
+        donor_data[donor_name] = [prompt_donation_amount(donor_name)]
+    else:
+        # Tried to make this for loop into a Comprehension but kept overwriting the current data instead of appending
+        # donor_data[donor_name] = [prompt_donation_amount(s) for s in donor_list if s.lower() == donor_name.lower()]
+        for s in donor_list:
+            if s.lower() == donor_name.lower():
+                # Update the donor's donation history
+                donor_data[s].append(prompt_donation_amount(donor_name))
```
[the remark this answers is not in the document]
Comment (rfelts) [reply students/rfelts/lesson05/mailroom.py]: Fixed in PR 113

File: students/rfelts/lesson05/mailroom.py [hunk 7]
```diff
@@ -0,0 +1,200 @@
...
+def create_donor_list():
+    """
+    Create a list of dicts containing: the donor name, last donation amount, and total donation amount.
+    :return: a list of donor dicts
+    """
+
+    donor_info = [{'name': name, 'last_donation': donations[-1], 'total_donations': sum(donations)}
+                  for name, donations in donor_data.items()]
+
+    return donor_info
```
[the remark this answers is not in the document]
Comment (rfelts) [reply students/rfelts/lesson05/mailroom.py]: Fixed in PR 113

File: students/rfelts/lesson05/mailroom.py [hunk 8]
```diff
@@ -0,0 +1,200 @@
...
+
+    # Create dynamic padding for the name
+    name_padding = len(max(donor_data, key=len)) + 4
+
+    # Build a list containing donor names, total donation amounts, number of donations, average donation
+    donors = list(donor_data.keys())
+    donor_values = list(donor_data.values())
+
+    temp_data = [[donor, sum(value[:]), len(value[:]), sum(value[:]) / len(value[:])]
+                 for donor, value in zip(donors, donor_values)]
```
[the remark this answers is not in the document]
Comment (rfelts) [reply students/rfelts/lesson05/mailroom.py]: Fixed in PR 113

File: students/rfelts/lesson05/mailroom.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (natasha-aleksandrova) [reply students/rfelts/lesson05/mailroom.py]: I see that now - you can keep it the way you have it.

**Pull request merged**